stream/stream_cb.c

```diff
@@ -0,0 +1,150 @@
+#include "config.h"
+
+#include <stdio.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <errno.h>
+
+#include "osdep/io.h"
+
+#include "common/common.h"
+#include "common/msg.h"
+#include "common/global.h"
+#include "stream.h"
+#include "options/m_option.h"
+#include "options/path.h"
+#include "player/client.h"
+#include "libmpv/stream_cb.h"
+
+struct mpv_stream_cb_context {
+    void *user_data;
+    mpv_stream_cb_open_fn open_fn;
+    mpv_stream_cb_read_fn read_fn;
+    mpv_stream_cb_seek_fn seek_fn;
+    mpv_stream_cb_size_fn size_fn;
+    mpv_stream_cb_close_fn close_fn;
+};
+
+mpv_stream_cb_context *mp_stream_cb_create(struct mpv_global *g,
+                                           struct mp_client_api *client_api)
+{
+    return talloc_zero(NULL, mpv_stream_cb_context);
+}
+
+void mpv_stream_cb_init(mpv_stream_cb_context *ctx, void *user_data)
+{
+    ctx->user_data = user_data;
+}
+
+void mpv_stream_cb_set_open_fn(mpv_stream_cb_context *ctx,
+                               mpv_stream_cb_open_fn fn)
+{
+    ctx->open_fn = fn;
+}
+
+void mpv_stream_cb_set_read_fn(mpv_stream_cb_context *ctx,
+                               mpv_stream_cb_read_fn fn)
+{
+    ctx->read_fn = fn;
+}
+
+void mpv_stream_cb_set_seek_fn(mpv_stream_cb_context *ctx,
+                               mpv_stream_cb_seek_fn fn)
+{
+    ctx->seek_fn = fn;
+}
+
+void mpv_stream_cb_set_size_fn(mpv_stream_cb_context *ctx,
+                               mpv_stream_cb_size_fn fn)
+{
+    ctx->size_fn = fn;
+}
+
+void mpv_stream_cb_set_close_fn(mpv_stream_cb_context *ctx,
+                                mpv_stream_cb_close_fn fn)
+{
+    ctx->close_fn = fn;
+}
+
+struct priv {
+    void *cookie;
+    mpv_stream_cb_context *stream_cb_ctx;
+};
+
+static int fill_buffer(stream_t *s, char *buffer, int max_len)
+{
+    struct priv *p = s->priv;
+    return (int)p->stream_cb_ctx->read_fn(p->cookie, buffer, (size_t)max_len);
+}
+
+static int seek(stream_t *s, int64_t newpos)
+{
+    struct priv *p = s->priv;
+    return (int)p->stream_cb_ctx->seek_fn(p->cookie, (off_t)newpos, SEEK_SET) != -1;
+}
+
+static int control(stream_t *s, int cmd, void *arg)
+{
+    struct priv *p = s->priv;
+    switch (cmd) {
+    case STREAM_CTRL_GET_SIZE: {
+        if (!p->stream_cb_ctx->size_fn)
+            break;
+        off_t size = p->stream_cb_ctx->size_fn(p->cookie);
+        if (size != (off_t)-1) {
```

Comment:
The off_t is apparently copy&pasted from stream_file.c, and doesn't make sense here.